File: app/src/main/java/com/chad/baserecyclerviewadapterhelper/activity/EmptyViewUseActivity.java
package com.chad.baserecyclerviewadapterhelper.activity;

import android.os.Bundle;
import android.view.View;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.chad.baserecyclerviewadapterhelper.R;
import com.chad.baserecyclerviewadapterhelper.adapter.EmptyViewAdapter;
import com.chad.baserecyclerviewadapterhelper.base.BaseActivity;
import com.chad.baserecyclerviewadapterhelper.data.DataServer;

public class EmptyViewUseActivity extends BaseActivity {

    private RecyclerView mRecyclerView;
    private final EmptyViewAdapter mAdapter = new EmptyViewAdapter();
    private boolean mError = true;
    private boolean mNoData = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_empty_view_use);

        setBackBtn();
        setTitle("EmptyView Use");

        findViewById(R.id.btn_reset).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                reset();
            }
        });

        mRecyclerView = findViewById(R.id.rv_list);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.setAdapter(mAdapter);
    }

    @Override
    protected void onStart() {
        super.onStart();
        onRefresh();
    }

    private void reset() {
        mError = true;
        mNoData = true;
        mAdapter.setList(null);
        onRefresh();
    }

    private View getEmptyDataView() {
        View notDataView = getLayoutInflater().inflate(R.layout.empty_view, mRecyclerView, false);
        notDataView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onRefresh();
            }
        });
        return notDataView;
    }

    private View getErrorView() {
        View errorView = getLayoutInflater().inflate(R.layout.error_view, mRecyclerView, false);
        errorView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onRefresh();
            }
        });
        return errorView;
    }

    private void onRefresh() {
        // 方式一：直接传入 layout id
        mAdapter.setEmptyView(R.layout.loading_view);
        mRecyclerView.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (mError) {
                    // 方式二：传入View
                    mAdapter.setEmptyView(getErrorView());
                    mError = false;
                } else {
                    if (mNoData) {
                        mAdapter.setEmptyView(getEmptyDataView());
                        mNoData = false;
                    } else {
                        mAdapter.setList(DataServer.getSampleData(10));
                    }
                }
            }
        }, 1000);
    }
}

File: app/src/main/java/com/chad/baserecyclerviewadapterhelper/activity/emptyview/EmptyViewUseActivity.kt
package com.chad.baserecyclerviewadapterhelper.activity.emptyview

import android.os.Bundle
import android.view.View
import android.widget.FrameLayout
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.chad.baserecyclerviewadapterhelper.R
import com.chad.baserecyclerviewadapterhelper.activity.emptyview.adapter.EmptyViewAdapter
import com.chad.baserecyclerviewadapterhelper.base.BaseViewBindingActivity
import com.chad.baserecyclerviewadapterhelper.data.DataServer
import com.chad.baserecyclerviewadapterhelper.databinding.ActivityEmptyViewUseBinding

class EmptyViewUseActivity : BaseViewBindingActivity<ActivityEmptyViewUseBinding>() {

    private val mAdapter = EmptyViewAdapter()

    private var mError = true
    private var mNoData = true

    override fun initBinding(): ActivityEmptyViewUseBinding =
        ActivityEmptyViewUseBinding.inflate(layoutInflater)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ViewCompat.setOnApplyWindowInsetsListener(viewBinding.titleBar) { view, insets ->
            val bar = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            view.updatePadding(top = bar.top)
            insets
        }

        viewBinding.titleBar.title = "EmptyView Use"
        viewBinding.titleBar.setOnBackListener { finish() }

        viewBinding.btnReset.setOnClickListener { reset() }

        viewBinding.rvList.adapter = mAdapter
        viewBinding.rvList.layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)

        // 打开空布局功能
        mAdapter.isStateViewEnable = true
        mAdapter.isUseStateViewSize = true

        onRefresh()
    }

    private fun reset() {
        mError = true
        mNoData = true
        mAdapter.submitList(null)
        onRefresh()
    }

    private val emptyDataView: View
        get() {
            val notDataView = layoutInflater.inflate(R.layout.empty_view, FrameLayout(this), false)
            notDataView.setOnClickListener { onRefresh() }
            return notDataView
        }

    private val errorView: View
        get() {
            val errorView = layoutInflater.inflate(R.layout.error_view, FrameLayout(this), false)
            errorView.setOnClickListener { onRefresh() }
            return errorView
        }

    private fun onRefresh() {
        // 方式一：直接传入 layout id
        mAdapter.setStateViewLayout(this, R.layout.loading_view)

        viewBinding.rvList.postDelayed({
            if (mError) { // 模拟网络错误
                // 方式二：传入View
                mAdapter.stateView = errorView

                mError = false
            } else {
                if (mNoData) { // 模拟接口没有数据
                    mAdapter.stateView = emptyDataView
                    mNoData = false
                } else {
                    // 模拟正常数据返回
                    mAdapter.submitList(DataServer.getSampleData(10))
                }
            }
        }, 1000)
    }
}